File: ios/Vxin/Features/Group/InviteMembersViewModel.swift
import Foundation

@MainActor
final class InviteMembersViewModel: ObservableObject {
    @Published var candidates: [Contact] = []
    @Published var selected: Set<String> = []
    @Published var loading = true
    @Published var inviting = false
    @Published var done = false
    @Published var error: String?

    let conversationId: String
    private let groupRepo = GroupRepository.shared
    private let contactRepo = ContactRepository.shared

    init(conversationId: String) {
        self.conversationId = conversationId
    }

    func load() async {
        loading = true
        do {
            let contacts = try await contactRepo.contacts()
            let memberIds = Set((try await groupRepo.info(conversationId)).members.map { $0.id })
            candidates = contacts.filter { !memberIds.contains($0.id) }
        } catch {
            self.error = (error as? LocalizedError)?.errorDescription ?? "加载失败"
        }
        loading = false
    }

    func toggle(_ id: String) {
        if selected.contains(id) { selected.remove(id) } else { selected.insert(id) }
    }

    func invite() {
        guard !selected.isEmpty, !inviting else { return }
        inviting = true; error = nil
        Task {
            do { try await groupRepo.invite(conversationId, userIds: Array(selected)); done = true }
            catch { self.error = (error as? LocalizedError)?.errorDescription ?? "邀请失败" }
            inviting = false
        }
    }
}

File: android/app/src/main/java/com/vxin/app/feature/group/InviteMembersViewModel.kt
package com.vxin.app.feature.group

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.vxin.app.core.network.toUserMessage
import com.vxin.app.data.model.Contact
import com.vxin.app.data.repository.ContactRepository
import com.vxin.app.data.repository.GroupRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class InviteMembersUiState(
    val loading: Boolean = true,
    val candidates: List<Contact> = emptyList(),   // 非群成员的联系人
    val selected: Set<String> = emptySet(),
    val inviting: Boolean = false,
    val done: Boolean = false,
    val error: String? = null,
)

@HiltViewModel
class InviteMembersViewModel @Inject constructor(
    private val groupRepository: GroupRepository,
    private val contactRepository: ContactRepository,
    savedStateHandle: SavedStateHandle,
) : ViewModel() {

    val conversationId: String = savedStateHandle.get<String>("conversationId").orEmpty()

    private val _uiState = MutableStateFlow(InviteMembersUiState())
    val uiState: StateFlow<InviteMembersUiState> = _uiState.asStateFlow()

    /** 一次性提示消费：Screen 展示 error 后调用，清空以免常驻 */
    fun consumeError() = _uiState.update { it.copy(error = null) }

    init {
        viewModelScope.launch {
            val result = runCatching {
                val contacts = contactRepository.contacts()
                val memberIds = groupRepository.info(conversationId).members.map { it.id }.toSet()
                contacts.filterNot { it.id in memberIds }
            }
            result
                .onSuccess { list -> _uiState.update { it.copy(loading = false, candidates = list) } }
                .onFailure { e -> _uiState.update { it.copy(loading = false, error = e.toUserMessage("加载失败")) } }
        }
    }

    fun toggle(id: String) = _uiState.update {
        it.copy(selected = if (id in it.selected) it.selected - id else it.selected + id)
    }

    fun invite() {
        val s = _uiState.value
        if (s.selected.isEmpty() || s.inviting) return
        _uiState.update { it.copy(inviting = true, error = null) }
        viewModelScope.launch {
            runCatching { groupRepository.invite(conversationId, s.selected.toList()) }
                .onSuccess { _uiState.update { it.copy(inviting = false, done = true) } }
                .onFailure { e -> _uiState.update { it.copy(inviting = false, error = e.toUserMessage("邀请失败")) } }
        }
    }
}
